fix core5compat error on windows

slspencer opened the pull request.

slspencer committed:
fix core5compat error on windows

ghost commented:
<details open>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=1017&src=https%3A%2F%2Fgithub.com%2FFashionFreedom%2FSeamly2D">Click on the image</a> for a new way to code review</b>
</summary>

####

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/FashionFreedom/Seamly2D/1017/a47c04a7/30d36287cdbc1c14bc528c7db051370ddd82b4d7.svg)](https://app.codesee.io/r/reviews?pr=1017&src=https%3A%2F%2Fgithub.com%2FFashionFreedom%2FSeamly2D)

#### Legend
<img src="https://app.codesee.io/Legend_v3.png" alt="CodeSee Map legend" width="220" height="156">
</details>

csett86 commented:
I may have a stupid question: Why do we need this? as far as I see it, it compiles fine on windows, both msvc and mingw without it...

slspencer commented:
@DSCaskey & I have had extensive problems compiling the qt6 branch on Windows with MSVC. I should've created an issue.
The compilation error was : "Unknown module in QT: core5compat" and the build would die. 
@DSCaskey said he didn't have the problem compiling with mingw. He tried a lot of tweaks, and can use MSVC now but it's unclear what he did to fix it.

csett86 commented:
The qt5compat module is not installed by default, so when installing qt, you need to make sure that you add it, same as QtMultimedia

Happy to enhance the developer README to make these things clear

slspencer commented:
That's just it, qt5 compatability was installed. I had it on my computer in the right directories.
I re-installed it several times using Qt Maintenance tool in case something had happened during installation.
"The qt5compat module is not installed by default, so when installing qt, you need to make sure that you add it, same as QtMultimedia"

It's a Qt6 problem. There are other reports of the core5compat not found error. Every suggested response is to install the qt 5 compatability module, but the person asking the question just lets the thread drop probably because this wasn't the fix. https://forum.qt.io/topic/124871/qt5compat-not-installed

@dscaskey said he could compile the qt6 branch on windows with msvc and qt 6.5.2 but only if he compiled qt from source.

csett86 commented:
Thats all very strange, and I really want to understand that. In the CI we also download Qt binaries from qt.io, same as the maintenancetool...

slspencer commented:
It's not a problem with the file download, because I had the files. I had to force QtCreator not to exclude the Qt 'include' lib directory (ironic). It's a problem with QtCreator. 

We don't use QtCreator in the CI -- we call qmake from the command line.

csett86 commented:
If its a problem with QtCreator, may I ask which version of QtCreator you used?

DSCaskey commented:
>@DSCaskey & I have had extensive problems compiling the qt6 branch on Windows with MSVC. I should've created an issue.

Not true for me. I had problems with the MinGw compiler as there was no prebuilt gcc lib for Xerces. Which is why I had to endure the pain of installing all the MSVC stuff. Once I had the kits in Creator configured I had no problem compiling Qt6 with MSVC 2019 64bit. What I did, and still do even after updating to Windows 11, is running the GCC 64 bit debugger. It tries to load a lib it can't register and fails. Switching to the MSVC 64 but kit and debugger fixed that, albeit if you turn on all warnings, it produced over 26,000 warnings (for MSVC libs), at which point I quit, and turned of all warnings. 

>@DSCaskey said he didn't have the problem compiling with mingw. He tried a lot of tweaks, and can use MSVC now but it's unclear what he did to fix it.

Didn't have to fix it (msvc)... just had to configure the Creator kits to find all the MSVC libs & kits. It was mingw that would not compile - as I said - there is no pre built Xerces libs for GCC. 

@DSCaskey said he could compile the qt6 branch on windows with msvc and qt 6.5.2 but only if he compiled qt from source.

Not even sure what that means? For 1 I haven't built yet with 6.5.2 as I just installed 6.5.2 with Web Assembly right before I broke my hip. Before that I could compile with 6.2.4 no problem.

DSCaskey commented:
>"The qt5compat module is not installed by default, so when installing qt, you need to make sure that you add it, same as QtMultimedia"

Right... when you use the Qt Maintence tool you need to make sure to check the  qt5compat module.

DSCaskey commented:
>It's not a problem with the file download, because I had the files. I had to force QtCreator not to exclude the Qt 'include' lib directory (ironic). It's a problem with QtCreator.

Creator does not "exclude" includes. It only looks for a Path variable  set in a kit's build environment to find modules... and that's assuming you have indeed installed the library module.

slspencer commented:
Wow, @dscaskey I misunderstood every single word you told me about your difficulties in compiling this qt6 fork.
It is essential that I verify that this qt6 fork can be compiled and that the resulting code works properly.   
@dscaskey Please provide the settings of your QtCreator compiler kit definition, Qt v6.4.2 is fine.   
Screen shots of your QtCreator kit tabs posted to this thread would be appreciated.

slspencer commented:
@csett86 - Here's my QtCreator version:
![image](https://github.com/FashionFreedom/Seamly2D/assets/578399/e27f473a-d304-4b84-8874-a4ff1efc74dc)

slspencer commented:
Here are screenshots of my QtCreator compiler kit settings on Windows using QtCreator 10.0.1, Qt 6.5.2, and MSVC 2019.
If someone is compiling this qt6 fork currently with similar settings then we know to look elsewhere for answers.

We need to find a documentable resolution for this problem. Any and all constructive answers are welcome.

slspencer commented:
@DSCaskey  > "when you use the Qt Maintenance tool you need to make sure to check the qt5compat module."

The qt5compat module was installed yet the "core5compat not found" errors occurred during compile.   
This problem has been [previously reported on the qt.io forum](https://forum.qt.io/topic/124871/qt5compat-not-installed). 

Here are the steps I took while troubleshooting, so that others can benefit:

- I verified in QtMaintenance tool that the qt5compat module box was already checked
- I verified that the qt5compat library was already installed in the correct directory.
- I updated & repaired Qt using the qtMaintenance Tool to force any missing or corrupt files to be downloaded & reinstalled.
- I cleaned & rebuilt Seamly, received the same core5compat errors so the compiler errored out and stopped.
-----
- I deleted Qt then reinstalled it, confirming again that the qt5compat module box was checked.
- Upon Seamly clean & rebuild, the same core5compat not found errors occurred & caused the compiler to stop.
-----
- I then thought about it some more.
- Hmmm. The qt5compat module was verifiably installed but the compiler was throwing 'not found' errors.
- The compiler wasn't finding it even though it's installed & where it should be . . .
- And wait . . . there's a compiler variable that defines which directories to search to find libraries.
- So I added the core5compat library directory to the compiler include_dir  variable.
- Upon Seamly clean & rebuild, the core5compat errors DID NOT APPEAR.
-----
- Sadly, upon running the Seamly debug version the code crashed -- This is the next issue we have to tackle after we solve & document the Windows qt6 development environment.

DSCaskey commented:
>@DSCaskey Please provide the settings of your QtCreator compiler kit definition, Qt v6.4.2 is fine.
Screen shots of your QtCreator kit tabs posted to this thread would be appreciated.

I already did in the discussion: https://github.com/orgs/FashionFreedom/discussions/996 albeit not with screencaps of all the edits to the system environment variables. 

![Screenshot (253)](https://github.com/FashionFreedom/Seamly2D/assets/31944718/d20f2fb4-eec0-40d4-8f14-a898f5fcfed1)

BTW... notice the Force UTF-8 MSVC checkbox... this was the easy solution to the whole UTF-8  char issue all along. ;)
